File: solutions/elixir/matching-brackets/4/lib/matching_brackets.ex
defmodule MatchingBrackets do
  @brackets %{"[" => "]", "{" => "}", "(" => ")"}
  @brackets_reversed Map.new(@brackets, fn {k, v} -> {v, k} end)
  @opening_brackets Map.keys(@brackets)
  @closing_brackets Map.values(@brackets)

  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t()) :: boolean
  def check_brackets(str) do
    str
    |> String.graphemes()
    |> Enum.reduce_while([], fn g, stack ->
      cond do
        g in @closing_brackets ->
          opening_bracket = Map.get(@brackets_reversed, g)
          if List.first(stack) === opening_bracket do
            {:cont, List.delete(stack, opening_bracket)}
          else
            {:halt, false}
          end
        g in @opening_brackets ->
          {:cont, [g | stack]}
        true ->
          {:cont, stack}
      end
    end)
    |> then(& &1 === [])
  end
end

File: solutions/go/matching-brackets/4/matching_brackets.go
package brackets

var (
	brackets         = map[rune]rune{'[': ']', '{': '}', '(': ')'}
	reversedBrackets = make(map[rune]rune, len(brackets))
)

func init() {
	for k, v := range brackets {
		reversedBrackets[v] = k
	}
}

func Bracket(input string) bool {
	var stack []rune

	for _, r := range input {
		if openingBracket, ok := reversedBrackets[r]; ok {
			if len(stack) == 0 || stack[0] != openingBracket {
				return false
			}
			stack = stack[1:]
		} else if _, ok := brackets[r]; ok {
			stack = append([]rune{r}, stack...)
		}
	}

	return len(stack) == 0
}
